botocore/auth.py

```diff
@@ -615,6 +628,285 @@ def add_auth(self, request):
         request.context['s3-presign-post-policy'] = policy
 
 
+class CrtSigV4Auth(BaseSigner):
+    REQUIRES_REGION = True
+    _PRESIGNED_HEADERS_BLACKLIST = [
+        'Authorization',
+        'X-Amz-Date',
+        'X-Amz-Content-SHA256',
+        'X-Amz-Security-Token',
+    ]
+    _SIGNATURE_TYPE = awscrt.auth.AwsSignatureType.HTTP_REQUEST_HEADERS
+    _USE_DOUBLE_URI_ENCODE = True
+    _SHOULD_NORMALIZE_URI_PATH = True
+
+    def __init__(self, credentials, service_name, region_name):
+        self.credentials = credentials
+        self._service_name = service_name
+        self._region_name = region_name
+        self._expiration_in_seconds = None
+
+    def _crt_request_from_aws_request(self, aws_request):
+        url_parts = urlsplit(aws_request.url)
+        crt_path = url_parts.path if url_parts.path else '/'
+        if aws_request.params:
+            array = []
+            for (param, value) in aws_request.params.items():
+                value = str(value)
+                array.append('%s=%s' % (param, value))
+            crt_path = crt_path + '?' + '&'.join(array)
+        elif url_parts.query:
+            crt_path = '%s?%s' % (crt_path, url_parts.query)
+
+        crt_headers = awscrt.http.HttpHeaders(aws_request.headers.items())
+
+        # CRT requires body (if it exists) to be an I/O stream.
+        crt_body_stream = None
+        if aws_request.body:
+            if hasattr(aws_request.body, 'seek'):
+                crt_body_stream = aws_request.body
+            else:
+                crt_body_stream = BytesIO(aws_request.body)
+
+        crt_request = awscrt.http.HttpRequest(
+            method=aws_request.method,
+            path=crt_path,
+            headers=crt_headers,
+            body_stream=crt_body_stream)
+        return crt_request
+
+    def _apply_signing_changes(self, aws_request, signed_crt_request):
+        # Apply changes from signed CRT request to the AWSRequest
+        aws_request.headers = HTTPHeaders.from_pairs(
+            list(signed_crt_request.headers))
+
+        if self._SIGNATURE_TYPE == \
+                awscrt.auth.AwsSignatureType.HTTP_REQUEST_QUERY_PARAMS:
+
+            new_query_string = urlsplit(signed_crt_request.path).query
+            p = urlsplit(aws_request.url)
+            # urlsplit() returns a tuple (and therefore immutable) so we
+            # need to create new url with the new query string.
+            # <part>   - <index>
+            # scheme   - 0
+            # netloc   - 1
+            # path     - 2
+            # query    - 3  <-- we're replacing this.
+            # fragment - 4
+            aws_request.url = urlunsplit(
+                (p[0], p[1], p[2], new_query_string, p[4]))
+
```

Comment:
Instead of putting an if statement specific to query signing in the main sigv4 signer, it may make sense to just override the `_apply_signing_changes()` method in the `CrtSigV4QueryAuth`. Otherwise, its a little hard to follow why this class has query signing logic and it would be nice to isolate all of the query specific logic to one class.